README
NAME
    UUID - DCE compatible Universally Unique Identifier library for Perl

SYNOPSIS
        use UUID ':all';

        generate($uuid);               # generate binary UUID, prefer random
        generate_random($uuid);        # generate binary UUID, using random
        generate_time($uuid);          # generate binary UUID, using time

        $string = uuid();              # generate stringified UUID

        unparse($uuid, $string);       # change $uuid string
        unparse_lower($uuid, $string); # change $uuid to lowercase string
        unparse_upper($uuid, $string); # change $uuid to uppercase string

        $rc = parse($string, $uuid);   # map string to UUID, return -1 on error

        copy($dst, $src);              # copy binary UUID from $src to $dst
        compare($uuid1, $uuid2);       # compare binary UUIDs

        clear( $uuid );                # set binary UUID to NULL
        is_null( $uuid);               # compare binary UUID to NULL

DESCRIPTION
    The UUID library is used to generate unique identifiers for objects that
    may be accessible beyond the local system. For instance, they could be
    used to generate unique HTTP cookies across multiple web servers without
    communication between the servers, and without fear of a name clash.

    The generated UUIDs can be reasonably expected to be unique within a
    system, and unique across all systems, and are compatible with those
    created by the Open Software Foundation (OSF) Distributed Computing
    Environment (DCE) utility uuidgen.

FUNCTIONS
    Most of the UUID functions expose the underlying libuuid C interface
    rather directly. That is, many return their values in their parameters
    and nothing else.

    Not very Perlish, is it? It's been like that for a long time though, so
    not very likely to change any time soon.

    All take or return UUIDs in either binary or string format. The string
    format resembles the following:

        1b4e28ba-2fa1-11d2-883f-0016d3cca427

    Or, in terms of printf(3) format:

        "%08x-%04x-%04x-%04x-%012x"

    The binary format is simply a packed 16 byte binary value.

  generate( $uuid )
    Generates a new binary UUID based on high quality randomness from
    /dev/urandom, if available.

    Alternately, the current time, the local ethernet MAC address (if
    available), and random data generated using a pseudo-random generator
    are used.

    The previous content of $uuid, if any, is lost.

  generate_random( $uuid )
    Generates a new binary UUID but forces the use of the all-random
    algorithm, even if a high-quality random number generator (i.e.,
    /dev/urandom) is not available, in which case a pseudo-random
    generator is used.

    Note that the use of a pseudo-random generator may compromise the
    uniqueness of UUIDs generated in this fashion.

  generate_time( $uuid )
    Generates a new binary UUID but forces the use of the alternative
    algorithm which uses the current time and the local ethernet MAC address
    (if available).

    This algorithm used to be the default one used to generate UUIDs, but
    because of the use of the ethernet MAC address, it can leak information
    about when and where the UUID was generated.

    This can cause privacy problems in some applications, so the generate()
    function only uses this algorithm if a high-quality source of randomness
    is not available.

  unparse( $uuid, $string )
    Converts the binary UUID in $uuid to string format and returns in
    $string. The previous content of $string, if any, is lost.

    The case of the hex digits returned may be upper or lower case, and is
    dependent on the system-dependent local default.

  unparse_lower( $uuid, $string )
    Same as unparse() but $string is forced to lower case.

  unparse_upper( $uuid, $string )
    Same as unparse() but $string is forced to upper case.

  $rc = parse( $string, $uuid )
    Converts the string format UUID in $string to binary and returns in
    $uuid. The previous content of $uuid, if any, is lost.

    Returns 0 on success and -1 on failure. Additionally on failure, the
    content of $uuid is unchanged.

  clear( $uuid )
    Sets $uuid equal to the value of the NULL UUID.

  is_null( $uuid )
    Compares the value of $uuid to the NULL UUID.

    Returns 1 if NULL, and 0 otherwise.

  copy( $dst, $src )
    Copies the binary $src UUID to $dst.

    If $src isn't a UUID, $dst is set to the NULL UUID.

  compare( $uuid1, $uuid2 )
    Compares two binary UUIDs.

    Returns an integer less than, equal to, or greater than zero if $uuid1
    is less than, equal to, or greater than $uuid2.

    However, if either operand is not a UUID, falls back to a simple string
    comparison returning similar values.

  $string = uuid()
    Creates a new string format UUID and returns it in a more Perlish way.

    Functionally the equivalent of calling generate() and then unparse(),
    but throwing away the intermediate binary UUID.

EXPORTS
    All functions may be imported in the usual manner, either individually
    or all at once using the ":all" tag.

TODO
    Need more tests and sanity checks.

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2014 by Rick Myers.

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)

    Details of this license can be found within the 'License' text file.

AUTHOR
    Current maintainer:

      Rick Myers <jrm@cpan.org>.

    Authors and/or previous maintainers:

      Lukas Zapletal <lzap@cpan.org>

      Joseph N. Hall <joseph.nathan.hall@gmail.com>

      Colin Faber <cfaber@clusterfs.com>

      Peter J. Braam <braam@mountainviewdata.com>

SEE ALSO
    uuid(3), uuid_clear(3), uuid_compare(3), uuid_copy(3), uuid_generate(3),
    uuid_is_null(3), uuid_parse(3), uuid_unparse(3), perl(1).



Changes
Revision history for Perl extension UUID.

0.23 Sun Dec 28 18:58:04 EST 2014
  - Reverse order of rpc.h and rpcdce.h in UUID.xs (rpc.h first)

0.22 Sun Dec 28 08:42:23 EST 2014
  - Ugh! Forgot to include rpcdce.h in UUID.xs

0.21 Sun Dec 28 08:35:23 EST 2014
  - Instead of unistd.h, try rpcdce.h on Win
  - Minor cleanup in test.pl

0.20 Sat Dec 27 19:27:28 EST 2014
  - Also include unistd.h when using Rpc.h (Win)
  - Fix pointer problem in -luuid search on Win
  - Fix minor header propagation problem in Makefile.PL

0.19 Fri Dec 26 17:09:12 EST 2014
  - Fix pointer confusion in do_uuid() on BSD [rt.cpan.org #101137]

0.18 Fri Dec 26 08:20:06 EST 2014
  - Allow for non-true/false return from uuid_from_string() on BSD
  - Remove unused dependency on Config.pm
  - Added yet more debug info for 'make test' time
  - Change -DPERL__UUID__UUID_BUF_SZ to -DPERL__UUID__STRUCT_SZ

0.17 Fri Dec 26 05:05:13 EST 2014
  - Removed two nested comments on BSD /* sigh */
  - Let e2fs machines use sv_cmp too (let's see if OSX Frankenbox works)
  - Changed warns in Makefile.PL to prints

0.16 Thu Dec 25 18:24:23 EST 2014
  - Typo on Windows/BSD
  - Trying sv_cmp() instead of uuid_compare() on RPC-based machines

0.15 Thu Dec 25 05:55:13 EST 2014
  - REALLY fix do_uuid() typo. Really.
  - Fix pointer problems on BSD
  - Removed uuid_copy() from BSD. How did that get there? =)
  - Added yet another test for OSX segfault!
  - Check and set -DPERL__UUID__UUID_BUF_SZ=?? in Makefile.PL
    (Should be the same everywhere, right? Right?)

0.14 Wed Dec 24 22:39:42 EST 2014
  - Fix typo in Windows do_uuid() for compiler fail on strawberry
  - Now checking for all header files instead of taking first
  - Added more detail for failed interface search
  - Added check for uuid struct size

0.13 Wed Dec 24 07:45:49 EST 2014
  - Split -luuid search into header file and interface type stages
  - Removed SvPV_nolen everywhere (buffer overflows)

0.12 Wed Dec 24 03:07:50 EST 2014
  - Added support for non-e2fs systems (BSD, OSX, Win)

0.11_01 Tue Dec 23 20:23:29 EST 2014
  - Modified test.pl to hopefully catch where MacOSX fails
  - Heavy mods to Makefile.PL to differentiate between interfaces
  - Makefile.PL *lots* more verbose
  - Lots of #ifdefs added in XS
  - First shot at code that works for BSD style interface

0.11 Mon Dec 22 19:03:09 EST 2014
  - Looks like I've managed to cut the failure reports down a little TOO
    far, so now splitting the search for -luuid into two separate
    phases. The first, which I don't want to deal with right now, tests
    for usability of the -luuid flag by itself. The second, tests for
    usability of individual functions. We'll deal with platforms that
    actually find libuuid first, then MAYBE revisit others.

0.10 Mon Dec 22 11:55:30 EST 2014
  - Added clear() and is_null()
  - Added copy() and compare()
  - Added unparse_lower() and unparse_upper()
  - Added generate_random() and generate_time()

0.09 Mon Dec 22 04:31:26 EST 2014
  - Drag test.pl into this century using Test.pm
  - Add postamble to Makefile to generate README
  - Overhaul POD
  - Fix MacOS warning "unsigned char* <--> char*"

0.08 Sun Dec 21 06:06:21 EST 2014
  - Added README with pod2text UUID.pm README
  - Switched libuuid search to Devel::CheckLib

0.07 Sat Dec 20 21:12:17 EST 2014
  - Updated Makefile.PL
  - Added search for -luuid at Makefile.PL time
  - Updated license and POD to Artistic 2.0

0.06 Thu Dec 18 08:01:44 EST 2014
  - Took over maintaining (Rick Myers - JRM)
  - Added uuid()

0.05 Fri Dec 14 20:00:00 GMT 2012
  - Took over maintaining (Lukas Zapletal - LZAP)
  - Version bump (no changes)
  - Releasing in the original location

0.04 Wed Jul 22 20:17:26 PDT 2009
  - Seems to be abandoned (again)
  - Bump version number and upload to PAUSE

0.03  Fri Jan 12 15:24:24 MST 2007
  - Added Artistic license
  - Took over maintaining (Colin Faber - CFABER)

0.02  Unknown
  - unknown changes

0.01  Thu Feb  8 06:07:59 2001
  - original version; created by h2xs 1.20 with options -A -n UUID


test.pl
use warnings;
use Test;
BEGIN { plan tests => 34 }
use UUID;


UUID::debug();

UUID::generate( $bin );
ok length $bin, 16;

UUID::generate_random( $bin );
ok length $bin, 16;

UUID::generate_time( $bin );
ok length $bin, 16;

UUID::unparse( $bin, $str );
ok $str, qr{^[-0-9a-f]+$}i;
$rc = UUID::parse( $str, $bin2 );
ok $rc, 0;
ok $bin, $bin2;

UUID::unparse_lower( $bin, $str );
UUID::unparse_upper( $bin, $str2 );
ok length( $str ), 36;
ok $str, qr/^[-a-f0-9]+$/;
ok lc( $str ), lc( $str2 );
ok $str ne $str2;

# content of uuid is unchanged if parse fails
UUID::generate( $bin );
$bin2 = $bin;
$str = 'Peter is a moose';
$rc = UUID::parse( $str, $bin );
ok $rc, -1;
ok $bin, $bin2;

UUID::generate( $bin );
$rc = UUID::is_null( $bin );
ok $rc, 0;

UUID::clear( $bin );
ok length( $bin ), 16;
ok UUID::is_null( $bin ), 1;

$bin = 'bogus value';
ok UUID::is_null( $bin ), 0; # != the null uuid, right?

$bin = '1234567890123456';
ok UUID::is_null( $bin ), 0; # still not null

# make sure compare operands sane
UUID::generate( $bin1 );
$bin2 = 'x';
ok abs(UUID::compare( $bin1, $bin2 )), 1;
ok abs(UUID::compare( $bin2, $bin1 )), 1;
$bin2 = 'some silly ridulously long string that couldnt possibly be a uuid';
ok abs(UUID::compare( $bin1, $bin2 )), 1;
ok abs(UUID::compare( $bin2, $bin1 )), 1;

# sane compare
$uuid=1;
UUID::generate( $uuid ); # this is wrong. dont want to fix it though.
ok 1;
$bin2 = '1234567890123456';
ok 1;
$tmp1 = UUID::compare( $bin1, $bin2 );
ok 1;
$tmp2 = UUID::compare( $bin2, $bin1 );
ok 1;
$tmp2 = -UUID::compare( $bin2, $bin1 );
ok $tmp1, $tmp2;
ok UUID::compare( $bin1, $bin2 ), -UUID::compare( $bin2, $bin1 );
$bin2 = $bin1;
ok UUID::compare( $bin1, $bin2 ), 0;

# make sure we get back a null if src isnt sane
$bin1 = 'x';
UUID::copy( $bin2, $bin1 );
ok UUID::is_null( $bin2 ), 1;
$bin1 = 'another really really really long sting';
UUID::copy( $bin2, $bin1 );
ok UUID::is_null( $bin2 );

# sane copy
UUID::generate( $bin1 );
$bin2 = '1234567890123456';
UUID::copy( $bin2, $bin1 );
ok UUID::compare( $bin1, $bin2 ), 0;

# make sure we get back the same scalar we passed in
$bin1 = '1234567890123456';
UUID::generate( $bin2 );
$save1 = \$bin2;
UUID::copy( $bin2, $bin1 );
$save2 = \$bin2;
ok $save1, $save2;
ok $$save1, $$save2;

$rc = UUID::uuid();
ok length($rc), 36;


exit 0;


UUID.pm
package UUID;

require 5.005;
use strict;
#use warnings;

require Exporter;
require DynaLoader;

use vars qw(@ISA %EXPORT_TAGS @EXPORT_OK $VERSION);
@ISA = qw(Exporter DynaLoader);

# This allows declaration       use UUID ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.

%EXPORT_TAGS = (
    'all' => [qw(
        &clear &compare &copy &generate &generate_random &generate_time
        &is_null &parse &unparse &unparse_lower &unparse_upper &uuid
    )],
);

@EXPORT_OK = ( @{$EXPORT_TAGS{'all'}} );

$VERSION = '0.23';

bootstrap UUID $VERSION;

# Preloaded methods go here.

1;
__END__

=head1 NAME

UUID - DCE compatible Universally Unique Identifier library for Perl

=head1 SYNOPSIS

    use UUID ':all';

    generate($uuid);               # generate binary UUID, prefer random
    generate_random($uuid);        # generate binary UUID, using random
    generate_time($uuid);          # generate binary UUID, using time

    $string = uuid();              # generate stringified UUID

    unparse($uuid, $string);       # change $uuid string
    unparse_lower($uuid, $string); # change $uuid to lowercase string
    unparse_upper($uuid, $string); # change $uuid to uppercase string

    $rc = parse($string, $uuid);   # map string to UUID, return -1 on error

    copy($dst, $src);              # copy binary UUID from $src to $dst
    compare($uuid1, $uuid2);       # compare binary UUIDs

    clear( $uuid );                # set binary UUID to NULL
    is_null( $uuid);               # compare binary UUID to NULL
    

=head1 DESCRIPTION

The UUID library is used to generate unique identifiers for objects that
may be accessible beyond the local system. For instance, they could be
used to generate unique HTTP cookies across multiple web servers without
communication between the servers, and without fear of a name clash.

The generated UUIDs can be reasonably expected to be unique within a
system, and unique across all systems, and are compatible with those
created by the Open Software Foundation (OSF) Distributed Computing
Environment (DCE) utility uuidgen.

=head1 FUNCTIONS

Most of the UUID functions expose the underlying I<libuuid> C interface
rather directly. That is, many return their values in their parameters
and nothing else.

Not very Perlish, is it? It's been like that for a long time though, so
not very likely to change any time soon.

All take or return UUIDs in either binary or string format. The string
format resembles the following:

    1b4e28ba-2fa1-11d2-883f-0016d3cca427

Or, in terms of printf(3) format:

    "%08x-%04x-%04x-%04x-%012x"

The binary format is simply a packed 16 byte binary value.

=head2 B<generate(> I<$uuid> B<)>

Generates a new binary UUID based on high quality randomness from
I</dev/urandom>, if available.

Alternately, the current time, the local ethernet MAC address (if
available), and random data generated using a pseudo-random generator
are used.

The previous content of I<$uuid>, if any, is lost.

=head2 B<generate_random(> I<$uuid> B<)>

Generates a new binary UUID but forces the use of the all-random
algorithm, even if a high-quality random number generator (i.e.,
I</dev/urandom>) is not available, in which case a pseudo-random
generator is used.

Note that the use of a pseudo-random generator may compromise the
uniqueness of UUIDs generated in this fashion.

=head2 B<generate_time(> I<$uuid> B<)>

Generates a new binary UUID but forces the use of the alternative
algorithm which uses the current time and the local ethernet MAC address
(if available).

This algorithm used to be the default one used to generate UUIDs, but
because of the use of the ethernet MAC address, it can leak information
about when and where the UUID was generated.

This can cause privacy problems in some applications, so the B<generate()>
function only uses this algorithm if a high-quality source of randomness
is not available.

=head2 B<unparse(> I<$uuid>B<,> I<$string> B<)>

Converts the binary UUID in I<$uuid> to string format and returns in
I<$string>. The previous content of I<$string>, if any, is lost.

The case of the hex digits returned may be upper or lower case, and is
dependent on the system-dependent local default.

=head2 B<unparse_lower(> I<$uuid>B<,> I<$string> B<)>

Same as B<unparse()> but I<$string> is forced to lower case.

=head2 B<unparse_upper(> I<$uuid>B<,> I<$string> B<)>

Same as B<unparse()> but I<$string> is forced to upper case.

=head2 B<$rc = parse(> I<$string>B<,> I<$uuid> B<)>

Converts the string format UUID in I<$string> to binary and returns in
I<$uuid>. The previous content of I<$uuid>, if any, is lost.

Returns 0 on success and -1 on failure. Additionally on failure, the
content of I<$uuid> is unchanged.

=head2 B<clear(> I<$uuid> B<)>

Sets I<$uuid> equal to the value of the NULL UUID.

=head2 B<is_null(> I<$uuid> B<)>

Compares the value of I<$uuid> to the NULL UUID.

Returns 1 if NULL, and 0 otherwise.

=head2 B<copy(> I<$dst>B<,> I<$src> B<)>

Copies the binary I<$src> UUID to I<$dst>.

If I<$src> isn't a UUID, I<$dst> is set to the NULL UUID.

=head2 B<compare(> I<$uuid1>B<,> I<$uuid2> B<)>

Compares two binary UUIDs.

Returns an integer less than, equal to, or greater than zero if
I<$uuid1> is less than, equal to, or greater than I<$uuid2>.

However, if either operand is not a UUID, falls back to a simple string
comparison returning similar values.

=head2 B<>I<$string> B<= uuid()>

Creates a new string format UUID and returns it in a more Perlish way.

Functionally the equivalent of calling B<generate()> and then B<unparse()>, but
throwing away the intermediate binary UUID.

=head1 EXPORTS

All functions may be imported in the usual manner, either individually
or all at once using the "I<:all>" tag.

=head1 TODO

Need more tests and sanity checks.

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2014 by Rick Myers.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

Details of this license can be found within the 'License' text file.

=head1 AUTHOR

Current maintainer:

  Rick Myers <jrm@cpan.org>.

Authors and/or previous maintainers:

  Lukas Zapletal <lzap@cpan.org>

  Joseph N. Hall <joseph.nathan.hall@gmail.com>

  Colin Faber <cfaber@clusterfs.com>

  Peter J. Braam <braam@mountainviewdata.com>

=head1 SEE ALSO

B<uuid(3)>, B<uuid_clear(3)>, B<uuid_compare(3)>, B<uuid_copy(3)>,
B<uuid_generate(3)>, B<uuid_is_null(3)>, B<uuid_parse(3)>,
B<uuid_unparse(3)>, B<perl(1)>.

=cut



Makefile.PL
use strict;
use warnings;
use Devel::CheckLib;
use ExtUtils::MakeMaker;


# ARRRRR!.. there are _at_least_ 3 different UUID interfaces:
# Free/NetBSD/MacOSX, Windows, and e2fsprogs.
# ... this is gonna suck!
print "#\n";


# if the compiler can't find it,
# no sense wasting our time...
print "# ===> Checking for -luuid\n";
check_lib_or_exit(
    lib   => 'uuid',
    debug => 1,
);
print "#\n";


my @hfiles;
for my $try (qw{ uuid/uuid.h uuid.h rpc.h }) {
    print "# ===> Checking for $try\n";
    eval { assert_lib(
        lib      => 'uuid',
        header   => $try,
        debug    => 1,
    )};
    unless ( $@ ) {
        push @hfiles, $try;
        print "# ===> Found!\n";
    }
    print "#\n";
}

die "No header file found. Can't continue.\n" unless @hfiles;


my $header;
my $defs = [];
my $libs = [];

while ( $header = shift @hfiles ) {
    try_rpc(  $header ) and last;
    try_e2fs( $header ) and last;
    try_win(  $header ) and last;
}

die "No interface found. Can't continue.\n" unless $header;


sub try_rpc {
    my $hdr = shift;
    print "# ===> Checking for RPC interface in $hdr\n";
    eval { assert_lib(
        lib      => 'uuid',
        header   => $hdr,
        debug    => 1,
        function => qq/
            char str[37];
            uuid_t u1,u2;
            int32_t r,s;
            uuid_create(&u1,&s);
            uuid_create_nil(&u2,&s);
            uuid_from_string(str,&u1,&s);
            uuid_to_string(&u1,&str,&s);
            r = uuid_compare(&u1,&u2,&s);
            r = uuid_equal(&u1,&u2,&s);
            r = uuid_is_nil(&u1,&s);
            return 0;
        /,
    )};
    if ($@) { my $L = length $@; print "# ===> Err($L): \"$@\"\n"; return 0 }
    push @$defs, '-DPERL__UUID__RPC_INT';
    push @$libs, '-luuid';
    print "# ===> Found!\n";
    print "#\n";
    return 1;
}

sub try_e2fs {
    my $hdr = shift;
    print "# ===> Checking for e2fs interface in $hdr\n";
    eval { assert_lib(
        lib      => 'uuid',
        header   => $hdr,
        debug    => 1,
        function => qq/
            int i;
            char s[37];
            uuid_t u,uu;
            uuid_generate(u);
            uuid_generate_random(u);
            uuid_generate_time(u);
            uuid_unparse(u,s);
            uuid_unparse_lower(u,s);
            uuid_unparse_upper(u,s);
            uuid_parse(s,u);
            uuid_clear(u);
            uuid_copy(uu,u);
            i = uuid_compare(u,uu);
            i = uuid_is_null(u);
            return 0;
        /,
    )};
    if ($@) { my $L = length $@; print "# ===> Err($L): \"$@\"\n"; return 0 }
    push @$defs, '-DPERL__UUID__E2FS_INT';
    push @$libs, '-luuid';
    print "# ===> Found!\n";
    print "#\n";
    return 1;
}

sub try_win {
    my $hdr = shift;
    print "# ===> Checking for Win interface in $hdr\n";
    eval { assert_lib(
        lib      => 'rpcrt4',
        header   => [ $hdr, 'rpcdce.h' ],
        debug    => 1,
        function => qq/
            int i;
            UUID u,uu;
            RPC_CSTR cs;
            RPC_STATUS st;
            unsigned char *s;
            i  = UuidCompare(&u,&uu,&st);
            st = UuidCreate(&u);
            st = UuidCreateNil(&uu);
            st = UuidCreateSequential(&uu);
            i  = UuidEqual(&u,&uu,&st);
            st = UuidFromString(cs,&u);
            i  = UuidIsNil(&u,&st);
            st = UuidToString(&u,&s);
            return 0;
        /,
    )};
    if ($@) { my $L = length $@; print "# ===> Err($L): \"$@\"\n"; return 0 }
    push @$defs, '-DPERL__UUID__WIN_INT';
    push @$libs, '-lrpcrt4';
    print "# ===> Found!\n";
    print "#\n";
    return 1;
}

sub get_struct_size {
    my ( $hdrs, $libs ) = @_;
    print "# ===> Checking binary size\n";
    ( my $lib = $libs->[0] ) =~ s/-l//;
    check_lib(
        lib      => $lib,
        header   => $hdrs,
        debug    => 1,
        function => qq/
            return sizeof(uuid_t);
        /,
    );
    my $rv = $? >> 8;
    print "# ===> $rv\n";
    print "#\n";
    return $rv;
}

my $size = get_struct_size( $header, $libs );
push @$defs, "-DPERL__UUID__STRUCT_SZ=$size";


($header = '-DPERL__UUID__'. uc $header ) =~ y{/.}{__};
push @$defs, $header;


print "# ===> Writing Makefile\n";

WriteMakefile1(
    'NAME'               => 'UUID',
    'AUTHOR'             => 'Rick Myers <jrm@cpan.org>',
    'VERSION_FROM'       => 'UUID.pm',
    'ABSTRACT_FROM'      => 'UUID.pm',
    'LICENSE'            => 'artistic_2',
    'MIN_PERL_VERSION'   => '5.005',
    'LIBS'               => $libs,
    'DEFINE'             => join(' ', @$defs), #$header,
    'INC'                => '',
    'PREREQ_PM'          => {},
    'CONFIGURE_REQUIRES' => {
        'Devel::CheckLib' => '1.02',
    },
    'TEST_REQUIRES' => {
        'Test' => 0,
    },
    #'META_MERGE'       => {
    #    'resources' => {
    #        #repository => 'URL to repository here',
    #    },
    #},
    #BUILD_REQUIRES => {
    #},
);

sub WriteMakefile1 {  #Written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params=@_;
    my $eumm_version=$ExtUtils::MakeMaker::VERSION;
    $eumm_version=eval $eumm_version;
    die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
    die "License not specified" if not exists $params{LICENSE};
    if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
        #EUMM 6.5502 has problems with BUILD_REQUIRES
        $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
        delete $params{BUILD_REQUIRES};
    }
    delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
    delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
    delete $params{META_MERGE} if $eumm_version < 6.46;
    delete $params{META_ADD} if $eumm_version < 6.46;
    delete $params{LICENSE} if $eumm_version < 6.31;
    delete $params{AUTHOR} if $] < 5.005;
    delete $params{ABSTRACT_FROM} if $] < 5.005;
    delete $params{BINARY_LOCATION} if $] < 5.005;

    WriteMakefile(%params);
}

# stay calm. don't blink.
# this is just for me :-)
sub MY::postamble {
    return <<EOP;
readme:
	pod2text UUID.pm README
	perl -i -pe's{\\*(\\S+)\\*}{\\1}g' README
EOP
}

